File: day04.py
"""Solution to day 4 of Advent of Code"""

from get_input import get_input
from string import hexdigits, digits


def valid(passport, checks):
    for field, check in checks.items():
        if check is None:  # Skip optional fields
            continue
        if field not in passport or not check(passport[field]):
            return False
    return True


def part1(passports):
    checks = {
        "byr": lambda _: True,
        "iyr": lambda _: True,
        "eyr": lambda _: True,
        "hgt": lambda _: True,
        "hcl": lambda _: True,
        "ecl": lambda _: True,
        "pid": lambda _: True,
        "cid": None,  # Optional field
    }
    return sum(1 for passport in passports if valid(passport, checks))


def part2(passports):

    def height(f):
        if f.endswith("in"):
            return 59 <= int(f[:-2]) <= 76
        elif f.endswith("cm"):
            return 150 <= int(f[:-2]) <= 193
        return False

    def hair(f):
        return f[0] == "#" and \
            len(f) == 7 and \
            all(d in set(list(hexdigits)) for d in f[1:])

    def eye_color(f):
        valid_eye_colors = ('amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth')
        return f in valid_eye_colors

    def pid(f):
        return len(f) == 9 and all(d in digits for d in f)

    checks = {
        "byr": lambda f: 1920 <= int(f) <= 2002,
        "iyr": lambda f: 2010 <= int(f) <= 2020,
        "eyr": lambda f: 2020 <= int(f) <= 2030,
        "hgt": height,
        "hcl": hair,
        "ecl": eye_color,
        "pid": pid,
        "cid": None,
    }
    return sum(1 for passport in passports if valid(passport, checks))


def parse(text):
    entries = []
    entry = {}
    text += "\n"
    for line in text.splitlines():
        line = line.strip()
        if line == "" and entry != {}:
            entries.append(entry)
            entry = {}
        else:
            entry.update(dict(field.split(':') for field in line.split(' ')))
    return entries


if __name__ == "__main__":
    LINES = parse(get_input(day=4, year=2020))
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day04.rs
use common::get_input;
use std::collections::{HashMap,HashSet};

type Input = Vec<Option<Passport>>;

#[derive(Debug)]
struct Passport {
    byr: String,
    iyr: String,
    eyr: String,
    hgt: String,
    hcl: String,
    ecl: String,
    pid: String,
    cid: Option<String>,
}

impl Passport {
    fn new(line: String) -> Option<Passport> {
        let mut fields = HashMap::new();
        for field in line.split(" ") {
            let index = field.find(":").unwrap();
            fields.insert(field[0..index].to_string(), field[index+1..].to_string());
        }
        Some(Passport {
            byr: fields.get("byr")?.to_string(),
            iyr: fields.get("iyr")?.to_string(),
            eyr: fields.get("eyr")?.to_string(),
            hgt: fields.get("hgt")?.to_string(),
            hcl: fields.get("hcl")?.to_string(),
            ecl: fields.get("ecl")?.to_string(),
            pid: fields.get("pid")?.to_string(),
            cid: fields.get("cid").map(|s| s.to_string()),
        })
    }

    fn is_valid(&self) -> bool {
        self.valid_hgt() && self.valid_byr() && self.valid_iyr() &&
            self.valid_hcl() &&self.valid_ecl() && self.valid_pid() && self.valid_eyr()
    }

    fn valid_byr(&self) -> bool {
        if let Ok(year) = self.byr.parse::<usize>() {
            return 1920 <= year && year <= 2002
        }
        false
    }

    fn valid_iyr(&self) -> bool {
        if let Ok(year) = self.iyr.parse::<usize>() {
            return 2010 <= year && year <= 2020
        }
        false
    }

    fn valid_eyr(&self) -> bool {
        if let Ok(year) = self.eyr.parse::<usize>() {
            return 2020 <= year && year <= 2030
        }
        false
    }

    fn valid_hgt(&self) -> bool {
        if self.hgt.ends_with("cm") {
            if let Ok(hgt) = self.hgt[..self.hgt.len()-2].parse::<usize>() {
                return 150 <= hgt && hgt <= 193;
            }
        } else if self.hgt.ends_with("in") {
            if let Ok(hgt) = self.hgt[..self.hgt.len()-2].parse::<usize>() {
                return 59 <= hgt && hgt <= 76;
            }
        }
        false
    }

    fn valid_hcl(&self) -> bool {
        if self.hcl.len() == 7 && &self.hcl[..1] == "#" {
            let letters = self.hcl[1..].chars().collect::<HashSet<char>>();
            let hex = "abcdef0123456789".chars().collect::<HashSet<char>>();
            return letters.is_subset(&hex)
        }
        false
    }

    fn valid_ecl(&self) -> bool {
        let valid_colors = vec!["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
            .iter()
            .map(|&s| s)
            .collect::<HashSet<&str>>();
        valid_colors.contains(&self.ecl.as_str())
    }

    fn valid_pid(&self) -> bool {
        if self.pid.len() == 9 {
            return self.pid.parse::<usize>().is_ok()
        }
        false
    }
}

fn part1(passports: &Input) -> isize {
    passports.iter().filter(|p| p.is_some()).count() as isize
}

fn part2(passports: &Input) -> isize {
    passports.iter().filter(|p| {
        if let Some(p) = p {
            return p.is_valid()
        }
        false
    }).count() as isize
}

fn parse(text: String) -> Input {
    let mut fields: Vec<String> = vec![];
    let mut results = vec![];
    let text = text + "\n";
    for line in text.split('\n') {
        if line == "" && fields.len() > 0 {
            results.push(Passport::new(fields.join(" ")));
            fields.clear();
        } else {
            fields.push(line.to_string());
        }
    }
    results
}

fn main() {
    let input = parse(get_input(04, 2020));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
